File: maxed.pl
#!/usr/bin/perl
sub maxed{
#               BEGINNING OF MAXED UNFOLDING ALGORITHM      
	my $j;
	my $i;
	my $k;
	my $m;
	my $junk;

# Write output file "input_data"

	open(DATA,">maxed_data/input_data");

	my $total_groups=$num_groups+1;
	print DATA "$num_det,$total_groups\n";

	for($i=0;$i<$num_det;$i++){
		my $ii=$i+1;
		print DATA "$ii,$bce[$i],$errbce[$i]\n";
	}
	for($i=0;$i<$num_groups;$i++){
		print DATA "$eend[$i],$spli[$i]\n";
	}
	print DATA "$eend[$num_groups],0\n";
	print DATA "2,3\n";
	print DATA "1,0.85\n";
	close DATA;

# Write output file "response_function"

	open(DATA,">maxed_data/response");
	print DATA "$num_det\n";
	print DATA "$total_groups\n";
	print DATA "cm**2\n";
	print DATA "$eend[0]\n";
	for($i=1;$i<$num_groups+1;$i++){
		print DATA "$eend[$i]";
		for($j=0;$j<$num_det;$j++){
			print DATA ",$mat[$i][$j]";
		}
		print DATA "\n";
	}
	close DATA;

	system "killall -9 sand2";
	system "killall -9 maxed";
	$rm_file="./tmp/notezy/rm.$$";
	open(OUT,">$rm_file");
	print OUT "killall -9 maxed\n";
    print OUT "rm $rm_file\n";
	close(OUT);
	print `at -f $rm_file now + 15 minutes`;

	print "<pre>";
	chdir "maxed_data";	
	system "rm OUT.OUT";		
	system "rm OUT.TBL";
	print "DEBUG: launching MAXED\n";		
	print  `/usr/local/bin/maxed`;
	print "</pre>";	
	sleep 5;

	open(IN,"OUT.OUT");
	$junk=<IN>;
	$junk=<IN>;
	for($i=0;$i<$num_groups;$i++){
		$junk=<IN>;
		my @data=split(' ',$junk);
		$spl[$i]=$data[4];
		$splstart[$i]=$data[3];
	}
	close IN;
	system "rm OUT.OUT";		
	system "rm OUT.TBL";		
	chdir "..";
}
1;

File: bums2/algorithms/maxed.py
#The following will apply the maxed unfolding method
import subprocess
from pathlib import Path
from typing import Tuple, List
import numpy as np
import contextlib

from bums2.core.config import Bums2Config
from bums2.FORTRAN_METHODS.MAXED.driver.maxed_main import MaxedPipeline

class maxed:
    def __init__(self, cfg: Bums2Config, workdir: Path = Path("maxed_data")):
        self.cfg = cfg
        self.e_end = self.cfg.e_end
        self.workdir = workdir
        self.workdir.mkdir(exist_ok=True)

    def maxed_unfold(
            self,
            mat: np.ndarray,
            bce: np.ndarray,
            errbce: np.ndarray,
            spli: np.ndarray,
            num_groups: float,
            num_det: float,
            out: Path = None
    ) -> Tuple[np.ndarray, np.ndarray]:
        #Perform MAXED unfolding
        #Parameters:
        #mat : ndarray, shape(num_detectors, num_groups), The response matrix
        #bce: ndarray, shape(num_detectors,), The measured counts
        #errbce: ndarray, shape(num_detectors,), The error of the measured counts
        #spli: ndarray, shape(num_groups,), The initial spectrum
        
        #Returns:
        #spl: ndarray, shape(num_groups,), The unfolded spectrum
        #splstart: ndarray, shape(num_groups,), The initial spectrum feed into maxed
        total_groups = num_groups+1

        #Frist: write input_data file
        inp = self.workdir / "input_data"
        with inp.open("w") as fh:
            fh.write(f"{num_det},{total_groups}\n")

            #For detector lines index out bce and errbce
            for i in range(num_det):
                fh.write(f"{i+1},{bce[i]},{errbce[i]}\n")

            #Parse the energy and initial spectrum values
            for j in range(num_groups):
                fh.write(f"{self.cfg.e_end[j]},{spli[j]}\n")
            fh.write(f"{self.cfg.e_end[num_groups]},0\n")

            #Some harcoded lines from the original perl file
            fh.write("2,3\n")
            fh.write("1,0.85\n")

        #Second: write a response file
        resp = self.workdir / "response"
        with resp.open("w") as fh:
            fh.write(f"{num_det}\n")
            fh.write(f"{total_groups}\n")
            fh.write(f"cm**2\n")

            #First energy endpoint
            fh.write(f"{self.cfg.e_end[0]}\n")

            #Form the response matrix where each line is eend[j],value
            for i in range(total_groups - 1):  # i = 0 to num_groups - 1
                fh.write(f"{self.cfg.e_end[i + 1]}")  # Upper bin edge, equivalent to Perl's eend[i]
                for j in range(num_det):
                    fh.write(f",{mat[i][j]}")
                fh.write("\n")
        
        #Third: Call the maxed pipeline and find spl and splstart
        pipeline = MaxedPipeline(
            input_file= inp,
            response_file= resp,
            iqds=2,
            iqbs=3,
        )
        
        with open(out, "a") as f, contextlib.redirect_stdout(f):
            result = pipeline.run()

        spl = result["FL"]
        splstart = result["FIL"]
      
        return spl, splstart